src/sisl/io/sile.py

```diff
@@ -722,72 +718,44 @@
-                    start = 0
-                if stop is None:
-                    stop = 100000000000
-                if step is None:
-                    step = 1
-            else:
-                # behave like range(stop) | range(start, stop, step)
-                # almost.
-                #  step only => start, step = 0, inf
-                #  start only => stop = start + 1
-                #  stop only => start = stop - 1
-                if step is not None:
-                    if start is None:
-                        start = 0
-                    if stop is None:
-                        stop = -1
-                elif start is None and stop is None:
-                    start = 0
-                    stop = 1
-
-                if start is None:
-                    if stop is not None:
-                        start = stop - 1
-                    else:
-                        start = 0
-                if stop is None:
-                    # this will only occur if start is given, but not stop
-                    stop = start + 1
-                elif stop < 0:
-                    stop = 1000000000000
-                if step is None:
-                    # only handle cases where the requested number of
-                    # elements is longer than 1, then step should be defined,
-                    # else we resort to returning a single entity (not postprocessed)
-                    if stop - start > 1:
-                        step = 1
-
-            if stop < 0:
-                # TODO this will probably fail if users do stop=-4 (to not have the last 3 items)
-                # It should be done in a different manner
-                stop = 100000000000
+            inf = 100000000000
 
             def check_none(r):
                 if isinstance(r, tuple):
                     return reduce(lambda x, y: x and y is None, r, True)
                 return r is None
 
-            # start by reading past start, this is regardless of what we
-            # are about to do
-            for _ in range(start):
-                r = skip_call(*args, **kwargs)
-                if check_none(r):
-                    return r
-
+            if start is None:
+                start = 0
+            if stop is None:
+                stop = inf
             if step is None:
-                # read a single element and return
-                def ret_func():
-                    return wrap_func(*args, **kwargs)
+                step = 1
 
+            if start < 0 or stop < 0 or step < 0:
+                # we need to read everything and take a slice afterwards
+                def ret_func():
+                    R = []
+                    for _ in range(inf):
+                        r = wrap_func(*args, **kwargs)
+                        if check_none(r):
+                            break
+                        R.append(r)
+                    slc = slice(start, stop, step)
+                    R = R[slc]
```

Comment:
I don't understand the error indicated here. Slicing a list should be perfectly fine, no?

Reply:
Maybe `R = R[...]`, i.e. naming the result with the same name as the list, messes with the way the algorithm determines types, and it assumes that `R` can be anything, not just a list (although I agree it seems it would be a misjudgement :smile: ).